### DISCUSSION TITLE
Uso de funciones TP1-EJ2

### DISCUSSION THREAD

**Discussion by Clirussa** [INGCOM-UNRN-P1/cursada-2023#82] [Preguntas y respuestas]
Hola, estamos viendo con @agosgit  como se implementan las funciones. Tenemos esto codificado, pero nos salta error en las líneas 14; 19; 36 y 43. No sabemos qué está mal. 
Esperamos sugerencias, muchas gracias.

![ejercicio2-tp1](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/102642611/c1b138da-2f30-4627-8471-2755724332e3)

**Comment by martinvilu** [comment]
¿Y que es lo que está mal?

¿Es algo de la corrección?, ¿es algo del compilador?

Si es el compilador, ¿cual es el mensaje?

**Reply by Clirussa** [reply]
Eso es lo que dice el mensaje. 

![ejercicio2-tp1](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/102642611/8fe86e98-647e-4826-8f59-7cc9310771a9)

**Reply by martinvilu** [reply]
¿En donde se declara la variable `resultado` que esta siendo utilizada en las funciones de conversión?

**Comment by Clirussa** [comment]
Más allá del ejercicio puntual, hay varias cosas que no nos quedan claras. 

-¿Se debe declarar la función y después se la escribe de nuevo y se escribe el bloque que debe ejecutarse?
```C
float centigrados_a_fahrenheit(float centigrados);

float fahrenheit_a_centigrados(float fahrenheit);

float centigrados_a_fahrenheit(float centigrados)
{
    resultado = (9/5 * centigrados) + 32;
}

float fahrenheit_a_centigrados(float fahrenheit)
{
        resultado = (5/9) * (fahrenheit - 32);
}   
```

¿Está bien así o se la define así directamente?
```C
float centigrados_a_fahrenheit(float centigrados)
{
    resultado = (9/5 * centigrados) + 32;
}
```

 * Cuando dice que hay error en la línea 19, tenemos entendido que las variables deben ser locales, pero aparece "resultado" como no definida. No entendemos porque está mal eso. 

 * Cuando capturamos el resultado del valor de la temperatura ingresado por el usuario, no lo está tomando al momento de que se ingrese a la fórmula para la conversión ¿Se debe hacer algo más? o ¿Se hace de otra forma?
 * ¿Los IF se ubican dentro de main o se pueden ubicar por fuera? ¿O depende de algo?

Esas son algunas de las dudas que tenemos, además de que no funcione el código.

**Reply by martinvilu** [reply]
> ¿Se debe declarar la función y después se la escribe de nuevo y se escribe el bloque que debe ejecutarse?

![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/6e0a509b-86a8-41ab-b34f-3c27b5dd19e4)

Los prototipos son necesarios cuando la función se llama antes de que pueda ser declarada.

Ambas formas son correctas, pero si ubicas el `main` antes que las mismas, no funcionará porque no son “conocidas” y ahí si es necesario declarar antes de implementar.

> Cuando dice que hay error en la línea 19, tenemos entendido que las variables deben ser locales, pero aparece "resultado" como no definida. No entendemos porque está mal eso.

Explicame esto con un poco más de detalle, por favor.

> Cuando capturamos el resultado del valor de la temperatura ingresado por el usuario, no lo está tomando al momento de que se ingrese a la fórmula para la conversión ¿Se debe hacer algo más? o ¿Se hace de otra forma?

El código no es correcto por un conjunto de situaciones, incluyendo la llamada a la función de conversión y su declaración.

> ¿Los IF se ubican dentro de main o se pueden ubicar por fuera? ¿O depende de algo?

Van donde sea necesario decidir algo.

**Reply by JeanFranco99** [reply]
Una función es un bloque de código dentro del programa que se encargan de realizar una tarea determinada fuera del main, siendo esta ultima la principal, como su nombre le indica. 
Existen dos tipos de funciones. Las que devuelven (o retornan) un valor, y las que no devuelven ninguno (void)
>float centigrados_a_fahrenheit(float centigrados)
>{
>resultado = (9/5 * centigrados) + 32;
>}

Lo anterior declara una funcion llamada centigrados_a_fahrenheit que regresa un valor float y acepta otro valor float como parametro. Al estar realizando una operacion dentro de la funcion debemos colocarle la instrucción return, con la cual devolverá un valor específico (en este caso la variable resultado, que se le asigno (9/5 * centigrados) + 32. 

> Cuando capturamos el resultado del valor de la temperatura ingresado por el usuario, no lo está tomando al momento de que se ingrese a la fórmula para la conversión ¿Se debe hacer algo más? o ¿Se hace de otra forma?

Cuando capturas con el scanf(siempre se hace dentro de la funcion principal, es decir del main()) tenes que hacer un llamado a la funcion que vas a ingresar ese parametro, lo recomendable es hacerlo con un variable "auxiliar"
Ej: 
float resultado_conversion_cent;    //declaro la variable auxiliar 
scanf("%f", &centigrados) //  se ingresa el valor, en este caso centigrados
resultado_conversion_cent = centigrados_a_fahrenheit(centigrados) //la misma variable que capturaste anteriormente, se la estas pasando como parametro a la funcion que realizaste.

>¿Los IF se ubican dentro de main o se pueden ubicar por fuera? ¿O depende de algo?

Depende para que necesites el IF. Se pueden ubicar en las funciones o en el main, segun sea el caso que se necesite

**Comment by martinvilu** [comment]
![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/13653a03-9c71-4930-adf7-28625ce4d123)


(1) La variable `resultado` no está declarada. (esto en Python funciona, pero en C no.)

(2) Por otro lado, esta función no “devuelve” ningún resultado.

(3) Declarar la función antes de implementarla es solo necesario cuando hay un lugar, como `main` que le de uso, en el medio.

**Comment by martinvilu** [comment]
Y sobre el `main`

![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/1f4e372d-3cdd-4b6a-b694-9fd7261e46fd)

(1) ¿En dónde guardas el resultado de la función que llamas?

(2) Esto declara una variable con el mismo nombre que el que usas para guardar el resultado de `scanf` por lo que la función no va a recibir nada.

**Reply by martinvilu** [reply]
Damos por cerrado este tema?
